src/tactic/positivity.lean

```diff
@@ -89,47 +102,153 @@ meta def positivity_canon : expr → tactic strictness
 
 namespace positivity
 
+/-- Inductive type recording whether the goal `positivity` is called on is nonnegativity, positivity
+or different from `0`. -/
+@[derive inhabited]
+inductive order_rel : Type
+| le  : order_rel -- `0 ≤ a`
+| lt  : order_rel -- `0 < a`
+| ne  : order_rel -- `a ≠ 0`
+| ne' : order_rel -- `0 ≠ a`
+
+meta instance : has_to_format order_rel :=
+⟨λ r, match r with
+  | order_rel.le := "order_rel.le"
+  | order_rel.lt := "order_rel.lt"
+  | order_rel.ne := "order_rel.ne"
+  | order_rel.ne' := "order_rel.ne'"
+  end⟩
+
 /-- Given two tactics whose result is `strictness`, report a `strictness`:
 - if at least one gives `positive`, report `positive` and one of the expressions giving a proof of
   positivity
-- if neither gives `pos` but at least one gives `nonnegative`, report `nonnegative` and one of the
+- if one reports `nonnegative` and the other reports `nonzero`, report `positive`
+- else if at least one reports `nonnegative`, report `nonnegative` and one of the
   expressions giving a proof of nonnegativity
+- else if at least one reports `nonzero`, report `nonzero` and one of the expressions giving a proof
+  of nonzeroness
 - if both fail, fail -/
 meta def orelse' (tac1 tac2 : tactic strictness) : tactic strictness := do
-  res ← try_core tac1,
-  match res with
+  res1 ← try_core tac1,
+  match res1 with
   | none := tac2
-  | some res@(nonnegative e) := tac2 <|> pure res
-  | some res@(positive _) := pure res
+  | some p1@(positive _) := pure p1
+  | some (nonnegative e1) := do
+      res2 ← try_core tac2,
+      match res2 with
+      | some p2@(positive _) := pure p2
+      | some (nonzero e2) := positive <$> mk_app ``lt_of_le_of_ne' [e1, e2]
+      | _ := pure (nonnegative e1)
+      end
+  | some (nonzero e1) := do
+      res2 ← try_core tac2,
+      match res2 with
+      | some p2@(positive _) := pure p2
+      | some (nonnegative e2) := positive <$> mk_app ``lt_of_le_of_ne' [e2, e1]
+      | _ := pure (nonzero e1)
+      end
   end
 
+infixr ` ≤|≥ `:2 := orelse'
```

Comment:
It's a cute notation but I'd argue this should be `localized` because I don't foresee applications outside of `positivity` itself.